### PR TITLE
SVG: Implement initial server-side support

### LINKED ISSUE #5474: SVG: Implement initial server-side support
## Feature Description

Following #5070 and https://github.com/google/web-stories-wp/issues/798#issuecomment-724891854, let's bring #5103 over the finish line.

Things to implement (non complete list):

* Tests
* Reject SVGs without viewBox
* SVG sanitization using 3rd party library
* Check to see if svgs are already supported. 

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Add server side support for the upload of svg files. 

## Relevant Technical Choices

This PR, enabled image/svg+xml files to be uploaded. It use XML parser to get height and width attributes ( and viewBox ), so save the height and width in post meta, in the format the WordPress requires. 

It also adds a check for script tags embedded in content of SVG. If SVG has a script tag and user doesn't have unfiltered uploads, the svg is rejected. 

I also cache the size of the svg in post meta. This will improve performance a little and keeps inline with a future core change. 

SVG parsing can be extensive, so it is cached in a array. In one request, a svg parsing can be requested 3 times. 

So SVGs are encoded as `image/svg`. This was pretty common in my testing. The correct type is `image/svg+xml`. I have added a workaround for this. 

As a temp workaround, I have added a const, that allows for svg uploads to be enabled in CLI so that SVG could be added for testing. 

I have added an svg in e2e test with a screenshot.

## To-do

- [x] Unit tests
- [x] Add check for XML ext is enabled. 

## Testing Instructions

Sample [data](https://dev.w3.org/SVG/tools/svgweb/samples/svg-files/). 
---

Fixes #5474

### CODE REVIEW

**Comment by swissspidy** [comment]
Was `assets/src/edit-story/icons/add.svg` deleted by accident?

File: tests/e2e/specs/editor/media/insertSVGImage.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ */
+import {
+  createNewStory,
+  clickButton,
+  withExperimentalFeatures,
+} from '../../../utils';
+
+const MODAL = '.media-modal';
+
+describe('Inserting SVG Image', () => {
```
Comment (swissspidy) [inline tests/e2e/specs/editor/media/insertSVGImage.js]: Do we need a test to actually assess that _uploading_ a new SVG file works?

File: includes/Plugin.php [hunk 2]
```diff
@@ -225,6 +225,10 @@ public function register() {
 		$this->analytics = new Analytics();
 		add_action( 'init', [ $this->analytics, 'init' ] );
 
+		// TODO, make this is a property. Not property as issue with phpmd.
```
Comment (swissspidy) [inline includes/Plugin.php]: There's now `@SuppressWarnings(PHPMD.TooManyFields)` in this class, so there should be no warnings anymore.

File: bin/local-env/install-wordpress.sh [hunk 3]
```diff
@@ -160,3 +160,7 @@ if [ "$WEBSTORIES_DEV_MODE" != $WEBSTORIES_DEV_MODE_CURRENT ]; then
   WEBSTORIES_DEV_MODE_RESULT=$(wp config get --type=constant --format=json WEBSTORIES_DEV_MODE | tr -d '\r')
   echo -e $(status_message "WEBSTORIES_DEV_MODE: $WEBSTORIES_DEV_MODE_RESULT...")
 fi
+
+wp config set WEBSTORIES_SVG_ENABLED true --raw --type=constant --quiet
+wp media import /var/www/html/wp-content/e2e-assets/video-play.svg
+wp config delete WEBSTORIES_SVG_ENABLED
```
Comment (swissspidy) [inline bin/local-env/install-wordpress.sh]: I don't think we need this whole constant just for the e2e tests, and I would like to avoid adding this constant if possible.\*

From what I can see this place here is the only reason. To make it work without a constant we could do this:

```bash
wp option patch insert web_stories_experiments enableSVG 1
wp media import /var/www/html/wp-content/e2e-assets/video-play.svg
wp option patch insert web_stories_experiments enableSVG 0
```

**Comment by spacedmonkey** [comment]
There is a workaround for this issue with the core library https://github.com/darylldoyle/svg-sanitizer/pull/42

File: tests/e2e/specs/editor/media/insertSVGImage.js (hunk 1 above)
Comment (spacedmonkey) [reply tests/e2e/specs/editor/media/insertSVGImage.js]: Does our test suite support uploading a file like that? I thought there was a reason we imported the images when the e2e test suite starts.

File: includes/SVG.php [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+	/**
+	 * Add SVG to allowed mime types.
+	 *
+	 * @param array $mime_types Associative array of allowed mime types per media type (image, audio, video).
+	 *
+	 * @return array
+	 */
+	public function web_stories_allowed_mime_types( array $mime_types ) {
+		$mime_types['image'][] = self::MINE_TYPE;
```
Comment (swissspidy) [inline includes/SVG.php]: MINE_TYPE -> MIME_TYPE

Comment (spacedmonkey) [reply includes/SVG.php]: Fixed in https://github.com/google/web-stories-wp/pull/5103/commits/20787b7b5c23d5e1060189dd0e28eed844f02542

File: package.json [hunk 5]
```diff
@@ -48,9 +48,12 @@
     "draft-js-import-html": "^1.4.1",
     "draftjs-filters": "^2.5.0",
     "flagged": "^2.0.1",
+    "fs": "0.0.1-security",
     "history": "^5.0.0",
     "mousetrap": "^1.6.5",
+    "os": "^0.1.1",
     "patch-package": "^6.2.2",
+    "path": "^0.12.7",
```
Comment (swissspidy) [inline package.json]: `fs`, `os`, `path` are built-in modules in node. They are not npm packages to install in `package.json`.

So need to undo these changes here.

File: tests/e2e/specs/editor/media/insertSVGImage.js (hunk 1 above)
Comment (swissspidy) [reply tests/e2e/specs/editor/media/insertSVGImage.js]: It should, yes. Puppeteer supports file uploads.

The reason for importing some images before running the test suite is so that we have some images available for testing, without having to upload them over and over again.

File: includes/SVG.php [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+		/**
+		 * Filter the check to see if svg support is already enabled.
+		 * This filter allows plugin developers to hotwire this check.
+		 *
+		 * @since 1.2.0
+		 *
+		 * @param bool $enabled Whether the svg support is already enabled.
+		 */
+		return apply_filters( 'web_stories_svg_already_enabled', $enabled );
```
Comment (swissspidy) [inline includes/SVG.php]: I can't think of a good use case for this filter. Could you perhaps elaborate?

File: includes/SVG.php [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+		add_filter( 'wp_check_filetype_and_ext', [ $this, 'wp_check_filetype_and_ext' ], 10, 5 );
+		add_filter( 'site_option_upload_filetypes', [ $this, 'filter_list_of_allowed_filetypes' ] );
+	}
+
+	/**
+	 * Helper function to check if svg uploads are already enabled.
+	 *
+	 * @return bool
+	 */
+	public function svg_already_enabled() {
```
Comment (swissspidy) [inline includes/SVG.php]: This should probably be private/protected.

File: includes/SVG.php [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+	public function upload_mimes_add_svg( array $mime_types ) {
+		// allow SVG file upload.
+		$mime_types['svg']  = self::MIME_TYPE;
+		$mime_types['svgz'] = self::MIME_TYPE;
+
+		return $mime_types;
+	}
+
+	/**
+	 * Add SVG to allowed mime types.
```
Comment (swissspidy) [inline includes/SVG.php]:
```suggestion
	 * Adds SVG to list of mime types and file extensions
```

Since `wp_get_mime_types()` does not deal with allowed mime types, just the list of available/existing ones.

File: includes/SVG.php [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+		$height     = isset( $attributes->height ) ? (int) $attributes->height : 0;
+
+		// If height and width are not set, try the viewport attribute.
+		if ( ! $width || ! $height ) {
+			$view_box = isset( $attributes->viewBox ) ? (string) $attributes->viewBox : '';
+			if ( empty( $view_box ) ) {
+				$view_box = isset( $attributes->viewbox ) ? (string) $attributes->viewbox : '';
+			}
+			$pieces = explode( ' ', $view_box );
+			if ( count( $pieces ) === 4 ) {
```
Comment (swissspidy) [inline includes/SVG.php]:
```suggestion
			if ( 4 === count( $pieces ) ) {
```

Although we're not so strict/consistent with yoda conditions in PHP code.

File: includes/SVG.php [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+				$width  = $pieces[2];
+				$height = $pieces[3];
```
Comment (swissspidy) [inline includes/SVG.php]:
```suggestion
				list (, , $width, $height ) = $pieces;
```

File: web-stories.php [hunk 11]
```diff
@@ -103,6 +103,8 @@ function web_stories_get_compat_instance() {
 		'libxml' => array(
 			'functions' => array(
 				'libxml_use_internal_errors',
+				'simplexml_load_string',
```
Comment (swissspidy) [inline web-stories.php]: `simplexml_load_string` is part of `simplexml`, not `libxml`.

So if we now require `simplexml` that should be its own separate check here, and also needs to be added to `composer.json` as a requirement (`ext-simplexml`).

However, I'd suggest looking into avoiding this new dependency if possible.

File: includes/SVG.php [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+		return $xml;
+	}
+
+	/**
+	 * Get SVG data.
+	 *
+	 * @param string $file File path.
+	 *
+	 * @return SimpleXMLElement
```
Comment (swissspidy) [inline includes/SVG.php]:
```suggestion
	 * @return string File contents.
```

File: includes/SVG.php [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+	 * Get xml document.
+	 *
+	 * @param string $svg String of xml.
+	 *
+	 * @return SimpleXMLElement|false
+	 */
+	protected function get_xml( $svg ) {
+		$errors = libxml_use_internal_errors( true );
+
+		$xml = simplexml_load_string( $svg );
```
Comment (swissspidy) [inline includes/SVG.php]: Let's try to avoid adding SimpleXML as a dependency.

Perhaps we can just use this?

```php
$doc = new DOMDocument();
$doc->loadXML( $svg );
```

File: scoper.inc.php [hunk 14]
```diff
@@ -142,6 +142,19 @@
...
+			->files()
+			->ignoreVCS( true )
+			->ignoreDotFiles( true )
+			->name( '*.php' )
+			->exclude(
+				[
+					'tests',
+				]
+			)
+			->in( 'vendor/enshrined/svg-sanitize' )
```
Comment (swissspidy) [inline scoper.inc.php]:
```suggestion
			->in( 'vendor/enshrined/svg-sanitize/src' )
```

Otherwise this includes `svg-scanner.php`, a small PHP command line utility that could be potentially dangerous to include in the plugin.

File: includes/SVG.php [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+	 * @param string $file File path.
+	 *
+	 * @return true|WP_Error
+	 */
+	protected function sanitize( $file ) {
+		$dirty     = $this->get_svg_data( $file );
+		$sanitizer = new Sanitizer();
+		$clean     = $sanitizer->sanitize( $dirty );
+
+		if ( false == $clean ) {
```
Comment (swissspidy) [inline includes/SVG.php]: Let's use `===`

File: includes/SVG.php [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+use Google\Web_Stories_Dependencies\enshrined\svgSanitize\Sanitizer;
+
+/**
+ * Class SVG
+ *
+ * @package Google\Web_Stories
+ */
+class SVG {
+	/**
+	 * File extenstion.
```
Comment (swissspidy) [inline includes/SVG.php]:
```suggestion
	 * File extension.
```

File: includes/SVG.php [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+	 * Mime type.
+	 *
+	 * @var string
+	 */
+	const MIME_TYPE = 'image/svg+xml';
+
+	/**
+	 * Array of SimpleXMLElements.
+	 *
+	 * @var array
```
Comment (swissspidy) [inline includes/SVG.php]: From what I've gathered it's an array of file contents.

```suggestion
	 * Cached list of SVG files and their contents.
	 *
	 * Speeds up access during the same request.
	 *
	 * @var string[]
```

File: includes/SVG.php [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+		$allowed_mime_types = get_allowed_mime_types();
+		$mime_types         = array_values( $allowed_mime_types );
+
+		$enabled = in_array( self::MIME_TYPE, $mime_types, true );
+
+		/**
+		 * Filter the check to see if svg support is already enabled.
+		 * This filter allows plugin developers to hotwire this check.
+		 *
+		 * @since 1.2.0
```
Comment (swissspidy) [inline includes/SVG.php]: Also need these for all the methods

File: tests/phpunit/tests/SVG.php [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+	use Private_Access;
+
+	public function setUp() {
+		parent::setUp();
+		add_filter( 'web_stories_svg_already_enabled', '__return_false' );
+	}
+
+	public function tearDown() {
+		parent::tearDown();
+		remove_filter( 'web_stories_svg_already_enabled', '__return_false' );
```
Comment (swissspidy) [inline tests/phpunit/tests/SVG.php]: The tests pass fine without hooking into this filter. Let's remove `setUp`/`tearDown` here.

File: tests/phpunit/tests/SVG.php [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+	public function test_upload_mimes_add_svg() {
+		$svg = $this->get_svg_object();
+		$svg->init();
+		$allowed_mime_types = wp_get_mime_types();
+		$mine_types         = array_values( $allowed_mime_types );
+		$this->assertContains( 'image/svg+xml', $mine_types );
+	}
+
+	/**
+	 * @covers ::upload_mimes_add_svg
```
Comment (swissspidy) [inline tests/phpunit/tests/SVG.php]:
```suggestion
	 * @covers ::mime_types_add_svg
```

File: tests/phpunit/tests/SVG.php [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\Tests;
+
+/**
+ * @coversDefaultClass \Google\Web_Stories\SVG
+ */
+class SVG extends \WP_UnitTestCase {
```
Comment (swissspidy) [inline tests/phpunit/tests/SVG.php:23]: There's no test for the `wp_handle_upload()` and `wp_check_filetype_and_ext()` methods

**Review by swissspidy — COMMENTED** [review]
Have yet to check out the tests more closely but so far it looks good  👍

File: tests/phpunit/tests/SVG.php (hunk 19 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/SVG.php]: Removed https://github.com/google/web-stories-wp/pull/5103/commits/4b6a406a34230deb3e0347b50e29363bc098cd90

File: includes/SVG.php (hunk 13 above)
Comment (spacedmonkey) [reply includes/SVG.php]: enshrined\svgSanitize library uses it, should I remove that as well?

File: includes/SVG.php (hunk 15 above)
Comment (spacedmonkey) [reply includes/SVG.php]:
```suggestion
		if ( false === $clean ) {
```

Comment (spacedmonkey) [reply includes/SVG.php]: I think this failed for some reason. Let's try again.

File: package.json (hunk 5 above)
Comment (spacedmonkey) [reply package.json]: Why did phpstorm ask me to install them....

Comment (spacedmonkey) [reply package.json]: Removed https://github.com/google/web-stories-wp/pull/5103/commits/9b29dd6b5613e35861a73016c67fab6c7ea70636

File: includes/SVG.php (hunk 13 above)
Comment (swissspidy) [reply includes/SVG.php]: I don't see any reference to SimpleXML at all in  that package. Where did you see that?

File: tests/phpunit/tests/SVG.php (hunk 21 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/SVG.php:23]: Tests added https://github.com/google/web-stories-wp/pull/5103/commits/7995a1ea92f043fb4c3103b7dcf8ec5aaffb291a

File: includes/SVG.php [hunk 22]
```diff
@@ -0,0 +1,428 @@
...
+	/**
+	 * Add SVG to allowed mime types.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param array $mime_types Associative array of allowed mime types per media type (image, audio, video).
+	 *
+	 * @return array
+	 */
+	public function web_stories_allowed_mime_types( array $mime_types ) {
```
Comment (swissspidy) [inline includes/SVG.php:176]: Nit: there's no test for this method.

Also, should it avoid duplicate entries? Perhaps doing `array_unique( $mime_types['image'] )`

File: includes/SVG.php [hunk 23]
```diff
@@ -0,0 +1,428 @@
...
+
+		// If height and width are not set, try the viewport attribute.
+		if ( ! $width || ! $height ) {
+			$view_box = $xml->getAttribute( 'viewBox' );
+			if ( empty( $view_box ) ) {
+				$view_box = $xml->getAttribute( 'viewbox' );
+			}
+			$pieces = explode( ' ', $view_box );
+			if ( 4 === count( $pieces ) ) {
+				list (, , $width, $height ) = $pieces;
```
Comment (swissspidy) [inline includes/SVG.php:307]: There is no coverage for this particular line. Don't we have a test SVG with a viewBox but no width/height?

File: includes/SVG.php [hunk 24]
```diff
@@ -0,0 +1,428 @@
...
+		}
+
+		$width  = (int) $xml->getAttribute( 'width' );
+		$height = (int) $xml->getAttribute( 'height' );
+
+		// If height and width are not set, try the viewport attribute.
+		if ( ! $width || ! $height ) {
+			$view_box = $xml->getAttribute( 'viewBox' );
+			if ( empty( $view_box ) ) {
+				$view_box = $xml->getAttribute( 'viewbox' );
```
Comment (swissspidy) [inline includes/SVG.php:303]: I thought only `viewBox` is valid? Should we even check for `viewbox`?

File: includes/SVG.php [hunk 25]
```diff
@@ -0,0 +1,428 @@
...
+		$loaded = $dom->loadXML( $svg );
+		if ( ! $loaded ) {
+			return false;
+		}
+		$node = $dom->getElementsByTagName( 'svg' )->item( 0 );
+		if ( ! $node ) {
+			return false;
+		}
+		libxml_clear_errors();
+		libxml_use_internal_errors( $errors );
```
Comment (swissspidy) [inline includes/SVG.php]: This is not run if returning `false` above, which means these settings won't be properly reset.

File: includes/SVG.php [hunk 26]
```diff
@@ -0,0 +1,428 @@
...
+	 */
+	protected function get_xml( $svg ) {
+		$dom                      = new DOMDocument();
+		$dom->preserveWhiteSpace  = false;
+		$dom->strictErrorChecking = false;
+
+		$errors = libxml_use_internal_errors( true );
+		$loaded = $dom->loadXML( $svg );
+		if ( ! $loaded ) {
+			return false;
```
Comment (swissspidy) [inline includes/SVG.php:395]: There's no test coverage for this line. Are we testing with an invalid XML file?

File: includes/SVG.php (hunk 22 above)
Comment (spacedmonkey) [reply includes/SVG.php:176]: I wasn't sure if there was value in testing a filter like this. But I can add a simple test.

File: includes/SVG.php (hunk 24 above)
Comment (spacedmonkey) [reply includes/SVG.php:303]: In my testing, I found many svg in the wild using this lower case. I think whatever generated these files made invalud svg. But either way we kind of have to support it.

File: includes/SVG.php (hunk 23 above)
Comment (spacedmonkey) [reply includes/SVG.php:307]: fixed in https://github.com/google/web-stories-wp/pull/5103/commits/7b55319d7d30890e2af1e0c2bab44c72af152e44

File: includes/SVG.php (hunk 26 above)
Comment (spacedmonkey) [reply includes/SVG.php:395]: Fixed https://github.com/google/web-stories-wp/pull/5103/commits/7b55319d7d30890e2af1e0c2bab44c72af152e44

**Review by swissspidy — COMMENTED** [review]
Looks like the sanitizer library isn't fully PHP8 compatible.

As per https://github.com/darylldoyle/svg-sanitizer/pull/45#issuecomment-730357918, there will be a new release shortly. In the meantime we could use the `master` branch or apply that patch manually.

**Comment by spacedmonkey** [comment]
> Looks like the sanitizer library isn't fully PHP8 compatible.
> 
> As per [darylldoyle/svg-sanitizer#45 (comment)](https://github.com/darylldoyle/svg-sanitizer/pull/45#issuecomment-730357918), there will be a new release shortly. In the meantime we could use the `master` branch or apply that patch manually.

Good catch. I have patched our library here https://github.com/google/web-stories-wp/pull/5103/commits/89cd3f381518df89adb1e9b976d086291ec297fb

**Comment by spacedmonkey** [comment]
I think this can land in 1.3 even if it is behind a flag.

**Review by swissspidy — APPROVED** [review]
